### PR TITLE
Internal Http SystemProxyInfo object should never be null

### PR DESCRIPTION
**Pull request by davidsh**
In order to provide a consistent experience with the HttpClient IWebProxy objects,
the returned internal proxy object which represent the system/platform proxy settings
should never be null. If the platform's settings indicate that no proxy is being used,
then return an instance of the internal HttpNoProxy object.

Note that even if the platform settings indicate that a proxy could be used, any
particular Http request might still not go thru a proxy. The final determination of
what proxy is being used for a request is still governed by the return of the
IWebProxy.IsBypassed and IWebProxy.GetProxy methods.

Contributes to #36553

### CODE REVIEW

**Comment by wfurt** [comment]
Why do you think this is better? Do you envision we would also do this for the new global static property?

**Comment by davidsh** [comment]
>Why do you think this is better? Do you envision we would also do this for the new global static property?

This required as part of implementing #36553.  We discussed this in the API review meeting for #36553.  We are exposing the platform/system proxy object for the HttpClient.DefaultProxy (static property).  Doing a 'get' of that property can never return null.  And doing a 'set' of the property doesn't allow a null value either.  See: #36553.

There are also some interesting implementation things I need to do when we expose HttpClient.DefaultProxy static property.  There is a relationship between that property and the instances properties of HttpClientHandler (UseProxy, Proxy) that have to work correctly with one another.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpNoProxy.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Net.Http
+{
+    internal sealed class HttpNoProxy : IWebProxy
+    {
+        public HttpNoProxy() {}
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpNoProxy.cs]: Nit: this shouldn't be necessary

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SystemProxyInfo.Windows.cs [hunk 2]
```diff
@@ -14,7 +14,7 @@ public static IWebProxy ConstructSystemProxy()
                 HttpSystemProxy.TryCreate(out proxy);
             }
 
-            return proxy;
+            return proxy ?? new HttpNoProxy();
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SystemProxyInfo.Windows.cs]: Should we instead check for null at the call site and create this HttpNoProxy there instead?

Comment (davidsh) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SystemProxyInfo.Windows.cs]: By "call site" do you mean SocketsHttpHandler which calls SystemProxyInfo.ConstructSystemProxy() ?

I don't want to do that.  I want ConstructSystemProxy() to always return a non-null IWebProxy object representing the current platform/system proxy settings.

And per my comment, https://github.com/dotnet/corefx/pull/37306#issuecomment-488171770, i will doing another PR to change this to a Singleton pattern (probably `Lazy<T>` for the return of the ConstructSystemProxy() method as part of finishing #36553

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SystemProxyInfo.Windows.cs]: > By "call site" do you mean SocketsHttpHandler which calls SystemProxyInfo.ConstructSystemProxy() ?

Yes.

> I don't want to do that. I want ConstructSystemProxy() to always return a non-null IWebProxy object representing the current platform/system proxy settings.

Ok.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpNoProxy.cs (hunk 1 above)
Comment (davidsh) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpNoProxy.cs]: I'll remove it. I need to remember that the compiler is smart.

**Pull request merged**